<launch>
  <!-- Basic parameters -->
  <param name="use_sim_time" type="bool" value="True"/>
  <!-- <arg name="rtabmap_args" default="-delete_db_on_start"/> -->
  <arg name="rtabmap_args" default="--database_path /home/fizzer/.ros/rtabmap.db"/>
  

  <!-- Static transform for base_footprint to RGB-D camera link (adjusted height) -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_camera_tf" args="0 0 0 1.570796327 3.141592654 0.5207963268 ground_truth/base_footprint camera_link 5" />
  
  <!-- RGB-D Mapping Configuration for RTAB-Map -->
  <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <!-- Set frame_id to ground_truth/base_footprint and odom_frame_id to ground_truth/world -->
      <param name="frame_id" type="string" value="ground_truth/base_footprint"/>
      <param name="odom_frame_id" type="string" value="ground_truth/world"/>
      <param name="wait_for_transform" type="bool" value="true"/>
      <param name="queue_size" type="int" value="100"/>
      <param name="approx_sync" type="bool" value="true"/>

      <!-- Enable Loop Closure and Mapping Features -->
      <param name="Mem/IncrementalMemory" type="bool" value="true"/>
      <param name="Mem/InitWMWithAllNodes" type="bool" value="true"/>
      <param name="RGBD/CreateOccupancyGrid" type="bool" value="true"/>
      <param name="RGBD/CloudMaxDepth" type="double" value="10.0"/>

      <!-- Robust optimization and localization adjustments -->
      <param name="Optimizer/Slam2D" type="bool" value="true"/>
      <param name="RGBD/OptimizeFromGraphEnd" type="bool" value="false"/>
      <param name="RGBD/OptimizeMaxError" type="double" value="0"/>
      <param name="Mem/DetectRate" type="double" value="0.5"/>
      <param name="RGBD/ProximityBySpace" type="bool" value="false"/>
      <param name="RGBD/ProximityMaxGraphDepth" type="int" value="0"/>
      <param name="Kp/MaxFeatures" type="int" value="1000"/>

      <param name="Grid/FromDepth" type="bool" value="true"/>
      <param name="Grid/RangeMax" type="double" value="5.0"/>
      <param name="Vis/MinInliers" type="int" value="1000"/>
      <param name="Vis/CorType" type="int" value="0"/>

      <!-- Use separate RGB and Depth topics -->
      <param name="subscribe_rgbd" type="bool" value="false"/>
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_rgb" type="bool" value="true"/>
      <param name="RGBD/Enabled" type="bool" value="true"/>

      <!-- Remap RGB and Depth topics -->
      <remap from="rgb/image" to="/B1/rgbd_camera/rgb/image_raw"/>
      <remap from="depth/image" to="/B1/rgbd_camera/depth/image_raw"/>
      <remap from="rgb/camera_info" to="/B1/rgbd_camera/rgb/camera_info"/>
      <remap from="depth/camera_info" to="/B1/depth_camera/depth/camera_info"/>
  </node>

  <!-- Visualization Configuration for RTAB-Map -->
  <node pkg="rtabmap_viz" type="rtabmap_viz" name="rtabmap_viz" output="screen">
    <param name="subscribe_depth" type="bool" value="true"/>
    <param name="frame_id" type="string" value="ground_truth/base_footprint"/>
    <param name="wait_for_transform" type="bool" value="true"/>
    <param name="queue_size" type="int" value="30"/>

    <!-- Remap RGB-D topics -->
    <remap from="rgb/image" to="/B1/rgb_camera/rgb/image_raw"/>
    <remap from="depth/image" to="/B1/depth_camera/depth/image_raw"/>
    <remap from="rgb/camera_info" to="/B1/rgb_camera/rgb/camera_info"/>
    <remap from="depth/camera_info" to="/B1/depth_camera/depth/camera_info"/>
    <remap from="/odom" to="/ground_truth/world"/>
  </node>
</launch>
